### DISCUSSION TITLE
No way to retrieve locale in middleware

### DISCUSSION THREAD

**Discussion by TommySorensen** [amannn/next-intl#256] [Q&A] [answered]
Moving suggestion from here https://github.com/amannn/next-intl/pull/149#issuecomment-1521742810

Currently it is not possible to retrieve the locale in the middleware meaning that you are left up to yourself if you need the locale for something else then routing.

On a project im working on, i need to know the locale in the middleware. I need it specifically because i have a function that rewrites the requested url to the nested route in the app folder. Currently the only way to do this in Next is to do it in the middleware since my site have language specific urls. So fx. `/en/localized-product-name/some-id-123` need to be rewrited to `/[locale]/[product]/[id]` route. I will do that by using this regex package also used internally by Next https://github.com/pillarjs/path-to-regexp

```js
const intlMiddleware = createMiddleware({
  locales: ['en', 'de'],
  defaultLocale: 'en',
});

export async function middleware(request) {
  const { locale, response } = intlMiddleware(request); // Something like this would work for me

  // Based on locale and request.nextUrl use path-to-regexp to rewrite to the nested route 
  // here, to do this i need the locale so i can fetch the CMS that has the logic of the urls

  return response;
}
```

@amannn FYI.

**Comment by amannn** [comment]
Hey @TommySorensen, thanks for moving the discussion here!

Is this about localizing pathnames then? There's a section in the docs about [localizing pathnames](https://next-intl-docs.vercel.app/docs/next-13/server-components#localizing-pathnames), did you have a look at this?

The docs link to the [named routes example](https://github.com/amannn/next-intl/tree/feat/next-13-rsc/packages/example-next-13-named-routes) which illustrates a pattern on how to solve this. There's also a dynamic parameter in the URL, that might be similar to your use case.

Let me know if this helps!

**Reply by TommySorensen** [reply]
I have yes, and that will work perfectly for most cases. But as soon as CMS editor change the url and publish, then you will need to trigger af build of the website since the rewrite config in the next.config file is only fetched at build time.

So basically what i need is the same as what you mentioned, i just need to do in the middleware by fetching the CMS for endpoints.
Performance is not an issue since i can use the caching functionality in next with fx. `caches.open('urls');` from the next `NextFetchEvent` to cache the data from the CMS for 1 minute or so.

**Reply by amannn** [reply, marked as answer]
I see! I'm curious how you implemented the programmatic rewrites in the middleware? I just experimented with this and found a way, but it's a bit a hack (see https://github.com/amannn/next-intl/issues/247#issuecomment-1523292342).

As for returning the `locale` from the middleware: This will likely not be possible, since I'd like the middleware to match the `NextMiddleware` type for simple usage and composition.

I'm considering built-in support for localized pathnames at some point, so I'd be curious to hear more about your use case and the patterns you're trying to support.

Here is a workaround how can currently extract the locale:

```tsx
import createMiddleware from 'next-intl/middleware';
import {NextFetchEvent, NextMiddleware, NextRequest} from 'next/server';

export default withExtraMiddleware(
  createMiddleware({
    locales: ['en', 'de'],
    defaultLocale: 'en'
  })
);

function withExtraMiddleware(next: NextMiddleware) {
  return async (request: NextRequest, event: NextFetchEvent) => {
    const response = await next(request, event);

    if (response && response.ok) {
      const url = new URL(
        response.headers.get('x-middleware-rewrite') || request.url
      );
      const locale = url.pathname.split('/')[1];
      console.log(locale);
    }

    return response;
  };
}

export const config = {
  // Skip all paths that should not be internationalized
  matcher: ['/((?!_next|.*\\..*).*)']
};

```

**Reply by TommySorensen** [reply]
Thx for the workaround. I will try to play with that 👍

To the question about rewrites you can boil it down to something like this
```js
const rewrites = await getRewrites(request, event);
const rewriteUrlMatch = Object.keys(rewrites).find(url => pathname.startsWith(url));

if (rewriteUrlMatch) {
  const url = request.nextUrl.clone();
  url.pathname = url.pathname.replace(rewriteUrlMatch, rewrites[rewriteUrlMatch]);

  return NextResponse.rewrite(url);
}
```

The getRewrites function would look similar to this:
```ts
const getRewrites = async (request: NextRequest, event: NextFetchEvent) => {
  const locale = 'en'; // Use workaround here to get locale
  const urlMappings = await getUrlMappings(locale, event);

  return {
    [urlMappings.productList]: `/${locale}/product`,
  };
};
```

`urlMappings.productList` would be something like `/someLocalizedProductListUrl` which essential is just a pointer to where the product list page is located in the CMS. So any url starting with this we now know is product list pages.

I have still not used the regex tester for adding `:id` or something like that. The above is just a simplified version of what i have running from a solution not using next-intl. This code is mostly just to get the idea of what i want to do

**Reply by TommySorensen** [reply]
In the example https://github.com/amannn/next-intl/issues/247#issuecomment-1526412338 @ryanburns23 is fetching the config in the middleware, meaning that the pathnames is now dynamically. But what will then happend with the `navigation.ts` file with the `createLocalizedPathnamesNavigation` function? That function can't be dynamic as i understand it, meaning that those 2 settings is kind of out of sync.

@amannn Would it be an issue if only the middleware gets the correct pathnames provided? I guess i will loose all the nice Typescript from the `createLocalizedPathnamesNavigation` function then.

I did try to use the `NextResponse.rewrite();` but could not get it to work correctly since the intl response is not attached to the response. I could not find a way to rewrite the response and still run the `createMiddleware` function on the request.

**Reply by amannn** [reply]
> Would it be an issue if only the middleware gets the correct pathnames provided? I guess i will loose all the nice Typescript from the createLocalizedPathnamesNavigation function then.

Theoretically you can, yes. And yes, the benefits of static typing go away here.

I guess the solution for you here really depends on your project. I have worked on one app where we use the `next-intl` middleware only for locale matching and then we have a single dynamic page like `[locale]/[[...slug]]` where localized pages from a CMS are fetched based on `locale` and `slug`. Currently I'm working on a more app-like app, where we use the static typing benefits for routes that are defined within the app.

There's probably no one-size-fits-all here, but based on your requirements you can use the parts that apply to your setup.

**Reply by TommySorensen** [reply]
I have the past 5 year used the catch-all approach you are mentioned, in both Angular and React/Next.js apps. And that have worked totally fine for when the CMS is in charge of the URL's. But there is so many cons that i'm done with it 🙂
The only solution there is to have localised URL's is to use the middleware to rewrite the URL's to the correct internal Next.js route. That will fix all the issues.

I did manage to find a way to do this with the `path-to-regexp` to detect if i need to rewrite the url. But as soon as i rewrite the url, then the page is just blank white with zero error message. 

So instead i will now try to do the rewrite in the `intlMiddleware` by fetching it "settings" in the middleware and just your API, since that should work out of the box.

This is some of the code to do the rewrite. Im also using the next-auth package, but i left out those parts here.

```js
import { compile, match, pathToRegexp } from 'path-to-regexp';

const lang = locales.join('|');
export const testUrls = {
  [`/:locale(${lang})?/some-other-url/:sku`]: `/:locale(${lang})?/product/:sku`,
};

function localizeUrls(request: NextRequest | NextRequestWithAuth, event: NextFetchEvent) {
  const shouldRewriteUrl = Object.entries(testUrls).find(([from, to]) => {
    const regexp = pathToRegexp(from);
    return regexp.test(request.nextUrl.pathname);
  });

  if (shouldRewriteUrl) {
    const [from, to] = shouldRewriteUrl;
    const fromPath = match(from)(request.nextUrl.pathname);
    const toPath = compile(to, { encode: encodeURIComponent });
    const url = request.nextUrl.clone();
    url.pathname = toPath(fromPath.params);
    
    console.log('rewrite to => ', url.pathname);
    
    const response = intlMiddleware(request);
    
    return NextResponse.rewrite(new URL(url.toString()), {
      request: { headers: response.headers },
    });
  }
  
  return intlMiddleware(request);
}
```
